exp/operator/controllers/shared.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cluster-api/api/v1alpha4"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func preflightChecks(ctx context.Context, c client.Client, provider client.Object) (*v1alpha4.Condition, ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(4).Info("Performing preflight checks.")
+
+	providerGVK := provider.GetObjectKind().GroupVersionKind()
+
+	providerList := &unstructured.UnstructuredList{}
```

Comment:
Wondering if we can avoid to work with unstructured, because this makes the code difficult to read and to maintain.
All the components now shares the same spec and status; could we leverage on this to implement shared functions?

Reply:
I don't have any ideas right now. Maybe an internal type for serializing providers?

Reply:
You are right we can use shared spec/status for such functions, but this specific case looks at objects' Kind and Name.

Reply:
What about about creating a GenericProvider type in the controller package with TypeMeta, ObjectMeta, Spec and Status.
Each controller should convert to this type, trigger a common reconcileNormal/reconcileDelete funcs and then convert back to the original type before patching.
We can eventually also add helpers method on GenericProvider like isCoreProvider etc etc.

Reply:
👍 I've just added it, ptal